Fix Issue 18011 - core.sys.condition cannot be used as shared

tom-tan opened the pull request:
It is to solve [Issue 18011](https://issues.dlang.org/show_bug.cgi?id=18011) by adding shared methods to `core.sync.condition.Condition`.
I also added a shared version of `Mutex#handleAddr` because it is used in `Condition#wait()` method.

thewilsonator commented:
Please make one of the commit messages say "Fix issue 18011: ..." so that the bot picks it up.

kubo39 commented:
@tom-tan 
These links could be helpful.
https://wiki.dlang.org/Starting_as_a_Contributor#How_to_choose_a_title_for_your_Pull_Request
https://wiki.dlang.org/Starting_as_a_Contributor#Squashing

CI failed due to style check, please make sure `make -f posix.mak style` success.

kubo39 commented:
This seems reasonable for me, but I'm not sure why these aren't marked as shared.
cc: @PetarKirov, because you are assigned at https://issues.dlang.org/show_bug.cgi?id=18011

MoonlightSentinel commented on src/core/sync/condition.d:
```diff
@@ -176,14 +201,27 @@ class Condition
+    }
+
+    /// ditto
+    void wait() shared
```
This overload could simply cast away shared and Call the existing wait method.
There is no need for a template.

Ditto for the other methods

tom-tan replied:
I updated to use casts instead of overloads except constructors.

tom-tan commented:
Thank you for valuable comments!

I update the title of this PR and the commit message.
I will add a test for shared later.

kubo39 commented:
L132: also needs to add `cast(pthread_cond_t*)`.

MoonlightSentinel commented on src/core/sync/condition.d:
```diff
@@ -75,6 +75,20 @@ class Condition
+    }
+
+    //
+    private this(this Q, M)( M m, bool _unused_ ) nothrow @safe
```
This could also be a plain method instead as it does not initialize const/immutable data AFAICT.

tom-tan replied:
It may cause a breaking change in ABI level.
In the case of using a plain method, we have to use an unsafe cast in `this(shared Mutex)` and have to change the attribute of `this(shared Mutex)` from `@safe` to `@trusted` but it changes ABI of the constructor (See [Type Mangling](https://dlang.org/spec/abi.html#type_mangling)).

But I am not sure it really happens because I am not familiar with low level programming.

PetarKirov replied:
@MoonlightSentinel Templating a member function on the type qualifier of `this` makes it possible to reuse code between `shared` and non-`shared` code without casts. This pattern is also used in `core.sync.mutex`.

MoonlightSentinel replied:
> It may cause a breaking change in ABI level.

Not really, it's a new symbol.

> In the case of using a plain method, we have to use an unsafe cast in `this(shared Mutex)` and have to change the attribute of `this(shared Mutex)` from `@safe` to `@trusted` [...]

You can always use a `@trusted` anonymous lamdba

MoonlightSentinel replied:
> @MoonlightSentinel Templating a member function on the type qualifier of `this` makes it possible to reuse code between `shared` and non-`shared` code without casts. This pattern is also used in `core.sync.mutex`.

Agreed, I was just thinking whether we need to have this kinda akward function header and  generate this constructor twice. 

But anyway, I'm fine if  this remains unchanged

MoonlightSentinel commented on src/core/sync/condition.d:
```diff
@@ -105,7 +119,7 @@ class Condition
                     if ( rc )
                         throw new SyncError( "Unable to initialize condition" );
-                    rc = pthread_cond_init( &m_hndl, &attr );
+                    rc = pthread_cond_init( cast(pthread_cond_t*)&m_hndl, &attr );
```
DStyle requires a space after a cast:

```suggestion
                    rc = pthread_cond_init( cast(pthread_cond_t*) &m_hndl, &attr );
```

Ditto below

MoonlightSentinel commented on src/core/sync/mutex.d:
```diff
@@ -296,6 +296,11 @@ package:
             return &m_hndl;
         }
+
+        shared(pthread_mutex_t*) handleAddr() shared
```
This seems unecessary?

PetarKirov replied:
Actually, does the non-`shared` `handleAddr` need to exist? It's marked as package, so I see no problem with also making it as `final nothrow @nogc`.

tom-tan replied:
If we fix this by using overloading such as `Mutex`, both versions are needed because they are called from `shared` and non-`shared` `wait()`, respectively.
If we fix it by using casting as @MoonlightSentinel suggested, only non-`shared` version is needed.

Which solution is more preferable?

tom-tan replied:
I will remove it because all the use of `handleAddr` is to be passed to `pthread_cond_*` that requires `pthread_mutex_t*`.
Casting `Condition#m_assocMutex` to `Mutex` seems better rather than providing an overload of `handleAddr` for `shared Mutex` because it is finally casted to non-`shared`.

tom-tan commented:
> L132: also needs to add cast(pthread_cond_t*).

Fixed!

I also added a unittest. It is almost the same as non-`shared` version but uses `shared Mutex` and `shared Condition` (I need comments for it because I am not good at low layer concurrent programming).

tom-tan commented:
One CI `buildkite/druntime` fails but this failure is not related to this request.

kubo39 commented:
> One CI tester buildkite/druntime fails but this failure is not related to this request.

Yes, this was already fixed in master at https://github.com/dlang/druntime/commit/5cd15b2dc4d3c2678d27aae8ab0f3900f9b93c1, but not merged into stable yet.

tom-tan commented:
Rebased!

tom-tan commented:
Now it passes the all CI tests!

tom-tan commented:
Thanks!

PetarKirov commented:
Thanks @tom-tan, nice work!
